src/insert_completer.hh

```diff
@@ -53,12 +53,13 @@ struct InsertCompletion
 {
     struct Candidate
     {
+        size_t input_order;
         String completion;
         String on_select;
         DisplayLine menu_entry;
 
         bool operator==(const Candidate& other) const { return completion == other.completion; }
-        auto operator<=>(const Candidate& other) const { return completion <=> other.completion; }
+        auto operator<=>(const Candidate& other) const { return std::tie(completion, input_order) <=> std::tie(other.completion, other.input_order); }
```

Comment:
This is a bit weird, equality and ordering are not consistent with each other.

Reply:
true. Happily we don't need this change at all because the comparison is only used for deduplicating full-line completions. For those I  don't think we care about preserving the input order.
And if we do care, we could use a stable sort for `<c-x>l` completions.

Same for buffer completions, I don't have a strong opinion on whether they should respect input order.
I'm leaning towards no but if we do want it it'd be

```c++
diff --git a/src/commands.cc b/src/commands.cc
index fb0b790dd..097895c98 100644
--- a/src/commands.cc
+++ b/src/commands.cc
@@ -185,8 +185,8 @@ static Completions complete_buffer_name(const Context& context, CompletionFlags
         if (RankedMatch match{bufname, query})
             matches.emplace_back(match, buffer.get());
     }
-    std::sort(filename_matches.begin(), filename_matches.end());
-    std::sort(matches.begin(), matches.end());
+    std::stable_sort(filename_matches.begin(), filename_matches.end());
+    std::stable_sort(matches.begin(), matches.end());
 
     CandidateList res;
     for (auto& match : filename_matches)
```